verboseTree-maven-plugin/src/main/java/org/apache/maven/dependency/graph/RepositoryUtility.java

```diff
@@ -0,0 +1,326 @@
+            ProjectBuildingRequest projectBuildingRequest = mavenExecutionRequest.getProjectBuildingRequest();
+
+            projectBuildingRequest.setRepositorySession( session );
+
+            // Profile activation needs properties such as JDK version
+            Properties properties = new Properties(); // allowing duplicate entries
+            properties.putAll( projectBuildingRequest.getSystemProperties() );
+            properties.putAll( OsProperties.detectOsProperties() );
+            properties.putAll( System.getProperties() );
+            projectBuildingRequest.setSystemProperties( properties );
+
+            ProjectBuilder projectBuilder = container.lookup( ProjectBuilder.class );
+            ProjectBuildingResult projectBuildingResult = projectBuilder.build( pomFile.toFile(),
+                    projectBuildingRequest );
+            return projectBuildingResult.getProject();
+        }
+        catch ( PlexusContainerException | ComponentLookupException | ProjectBuildingException ex )
+        {
+            throw new MavenRepositoryException( ex );
+        }
+    }
+
+    /**
+     * Returns Maven repository specified as {@code mavenRepositoryUrl}, after validating the syntax of the URL.
+     *
+     * @throws IllegalArgumentException if the URL is malformed for a Maven repository
+     */
+    public static RemoteRepository mavenRepositoryFromUrl( String mavenRepositoryUrl )
+    {
+        try
+        {
+            // Because the protocol is not an empty string (checked below), this URI is absolute.
+            new URI( mavenRepositoryUrl );
+        }
+        catch ( URISyntaxException ex )
+        {
+            throw new IllegalArgumentException( "Invalid URL syntax: " + mavenRepositoryUrl );
+        }
+
+        RemoteRepository repository = new RemoteRepository.Builder( null, "default", mavenRepositoryUrl ).build();
+
+        return repository;
+    }
+
+    private static VersionRangeResult findVersionRange( RepositorySystem repositorySystem,
+                                                        RepositorySystemSession session,
+                                                        String groupId, String artifactId )
+            throws MavenRepositoryException
+    {
+
+        Artifact artifactWithVersionRange = new DefaultArtifact( groupId, artifactId, null, "(0,]" );
+        VersionRangeRequest request = new VersionRangeRequest( artifactWithVersionRange,
+                Arrays.asList( RepositoryUtility.CENTRAL ), null );
+
+        try
+        {
+            return repositorySystem.resolveVersionRange( session, request );
+        }
+        catch ( VersionRangeResolutionException ex )
+        {
+            throw new MavenRepositoryException( ex );
+        }
+    }
+
+    /**
+     * Returns the highest version for {@code groupId:artifactId} in {@code repositorySystem}.
+     */
+    static String findHighestVersion( RepositorySystem repositorySystem, RepositorySystemSession session,
```

Comment:
I would expect this to return the version object and the caller to apply toString.